packages/components/src/components/VirtualizedList/VirtualizedList.tsx

```diff
@@ -48,133 +48,179 @@ const Item = styled.div`
     width: 100%;
 `;
 
-const calculateItemHeight = <T extends { height: number }>(item: T): number => item.height;
+type BaseItemProps = {
+    height: number;
+};
+
+const calculateItemHeight = <T extends BaseItemProps>(item: T): number => item.height;
+
+interface ListContainerProps<T extends BaseItemProps> {
+    listHeight: number | string;
+    listMinHeight: number | string;
+    totalHeight: number;
+    items: Array<T>;
+    itemHeights: Array<number>;
+    startIndex: number;
+    endIndex: number;
+    ref?: React.Ref<HTMLDivElement>; // NOTE: needs to be here due to typecasting due to forwardRef
+    renderItem: (item: T, index: number) => React.ReactNode;
+}
+
+function ListContainerComponent<T extends BaseItemProps>(
+    {
+        listHeight,
+        listMinHeight,
+        totalHeight,
+        items,
+        itemHeights,
+        startIndex,
+        endIndex,
+        renderItem,
+    }: ListContainerProps<T>,
+    ref: React.Ref<HTMLDivElement>,
+) {
+    return (
+        <Container ref={ref} $height={listHeight} $minHeight={listMinHeight}>
+            <Content style={{ height: `${totalHeight}px` }}>
+                {itemHeights.slice(startIndex, endIndex).map((height, index) => {
+                    const itemIndex = startIndex + index;
+                    const itemTop = itemHeights.slice(0, itemIndex).reduce((acc, h) => acc + h, 0);
+
+                    if (!items[itemIndex]) return null;
+
+                    return (
+                        <Item
+                            key={itemIndex}
+                            style={{
+                                top: `${itemTop}px`,
+                                height,
+                            }}
+                        >
+                            {renderItem(items[itemIndex], itemIndex)}
+                        </Item>
+                    );
+                })}
+            </Content>
+        </Container>
+    );
+}
 
-type VirtualizedListProps<T> = {
-    items: Array<T & { height: number }>;
+// NOTE: don't forget the forwardRef() because of passing the ref for useShadow()
+const ListContainer = memo(forwardRef(ListContainerComponent)) as typeof ListContainerComponent;
+
+type VirtualizedListProps<T extends BaseItemProps> = {
+    items: Array<T>;
     onScroll?: (e: Event) => void;
     onScrollEnd: () => void;
     listHeight: number | string;
     listMinHeight: number | string;
+    ref?: React.Ref<HTMLDivElement>;
```

Comment:
Unfortunatelly, due to typecast, this needs to be here to prevent TS errors.

Reply:
Oh yeah, `forwardRef` will be deprecated https://react.dev/reference/react/forwardRef

Reply:
yeah, this could have been better solution.